### PR TITLE
[ws-manager] Make cluster selection filter by application cluster

### PR DESCRIPTION
**Pull request by andrew-farries**
## Description

Redo of #14000 (reverted in https://github.com/gitpod-io/gitpod/pull/14029). The original lacked the changes to `components/gitpod-db/src/typeorm/workspace-cluster-db-impl.ts`. This PR adds that filtering and adds tests for it.

<details>
<summary>Context </summary>
As part of #9198  we want to start syncing the `d_b_workspace_cluster` table with `db-sync`.

Currently, the table differs between US and EU regions because each table contains only the data relevant to that region. For example, in the EU table, the `eu70` workspace cluster is marked as `available` and the `us70` cluster is `cordoned`.  In the US cluster `eu70` is `cordoned` and `us70` is `available`.

In order to sync the table we need to get to a point where there is no difference in the data in the table between EU and US regions.

To do that we will introduce a new field in the table called `applicationCluster` which records the name of the application cluster to which the record belongs. Thus, for each workspace cluster there will be two rows in Gitpod SaaS:

| name | applicationCluster | url | tls | state | ... |
| --- | --- | --- | --- | --- | --- |
| eu70 | eu02 | url | tls info | available | ... |
| eu70 | us02 | url | tls info | cordoned | ... |

Effectively the new `applicationCluster` column gives the table an extra dimension so that we can combine both tables (EU and US) into one.

#13722  added the column to the table and made `gpctl` fill the value when `gpctl register`ing a new workspace cluster. The value is taken from the `GITPOD_INSTALLATION_SHORTNAME` environment variable in `ws-manager-bridge`.
</details>

Make the cluster selection mechanism in `ws-manager-api` aware of the application cluster to which each workspace cluster is registered. Previously, the selection was by workspace cluster name only. Soon, as described in the **context** section, each region in gitpod will store records in the database for *all* workspace clusters - not just the ones in its region. This PR ensures that workspace clusters not in the same region as the application cluster will not be considered.

## Related Issue(s)

Part of https://github.com/gitpod-io/gitpod/issues/9198 and https://github.com/gitpod-io/gitpod/issues/13800

## How to test

Edit the `server` deployment and change the `WSMAN_CFG_MANAGERS` environment variable to change the `applicationCluster` to which the default workspace cluster in the preview environment is registered:

```json
[
  {
    "name": "",
    "url": "dns:///ws-manager:8080",
    "tls": {
      "ca": "/ws-manager-client-tls-certs/ca.crt",
      "crt": "/ws-manager-client-tls-certs/tls.crt",
      "key": "/ws-manager-client-tls-certs/tls.key"
    },
    "state": "available",
    "maxScore": 100,
    "score": 50,
    "govern": true,
    "admissionConstraints": null,
    "applicationCluster": "dev" // <- change this from "dev" to anything else
  }
]
```
The `WSMAN_CFG_MANAGERS` environment variable is base64 encoded.

Once the `server` deployment is edited it should no longer be possible to start a workspace as the only workspace cluster is now associated with a different (non-existent) application cluster:

image

Edit the `WSMAN_CFG_MANAGERS` environment variable once more and set the `applicationCluster` field back to `"dev"`. Starting a workspace should now proceed as normal.

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

File: components/gitpod-protocol/src/workspace-cluster.ts
Comment (geropl) [inline components/gitpod-protocol/src/workspace-cluster.ts:107]: To make `findFiltered` safer, we should make `applicationCluster` mandatory:
```
    findFiltered(predicate: WorkspaceClusterFilter): Promise<WorkspaceClusterWoTLS[]>;
...
export type WorkspaceClusterFilter = Pick<WorkspaceCluster, "applicationCluster"> & DeepPartial<Pick<WorkspaceCluster, "name" | "state" | "govern" | "url"> & { minScore: number }>
```
(or similar)

File: components/gitpod-protocol/src/workspace-cluster.ts
Comment (geropl) [inline components/gitpod-protocol/src/workspace-cluster.ts:104]: WE should definitely adjust `deleteBy` and `findBy` as well, and make them take a composite of the shape `Pick<WorkspaceCluster, "name" | "applicationCluster">`

File: components/ws-manager-api/typescript/src/client-provider-source.ts
Comment (geropl) [inline components/ws-manager-api/typescript/src/client-provider-source.ts:20]: Instead of reading from the env across the onject hiearchy let's trx to consume from `config.ts`. I think `Configuration.installation` is what we want.

File: components/ws-manager-api/typescript/src/client-provider-source.ts
Comment (geropl) [inline components/ws-manager-api/typescript/src/client-provider-source.ts:66]: same here.

File: components/ws-manager-api/typescript/src/client-provider-source.ts
Comment (geropl) [inline components/ws-manager-api/typescript/src/client-provider-source.ts:69]: Let's stick with `findById` :+1:

File: components/ws-manager-api/typescript/src/client-provider-source.ts
Comment (andrew-farries) [reply components/ws-manager-api/typescript/src/client-provider-source.ts:20]: This is a good point. I'm not comfortable with reading the environment here either. I'm trying a different DI-based approach at the moment where the applicationCluster is becomes part of `WorkspaceManagerClientProvider` and is set up in the DI container.

Branch is here: https://github.com/gitpod-io/gitpod/compare/af/di

**Comment by andrew-farries** [comment]
I plan to close this draft in favour of a series of smaller PRs using a DI approach, stating with the dependency injection PR: https://github.com/gitpod-io/gitpod/pull/14084

I'll address your comments here in each of those PRs @geropl.

**Comment by geropl** [comment]
@andrew-farries Why not pass the `installationName` in parallel to `name`?

**Comment by andrew-farries** [comment]
> @andrew-farries Why not pass the `applicationCluster` in parallel to `name`? Then there is no need for DI, at all, no?

@geropl The approach we synced about is implemented in 37eb088; passing the application cluster explicitly everywhere rather than injecting a new dependency into `WorkspaceManagerClientProvider` (as in https://github.com/gitpod-io/gitpod/pull/14084).

This approach results in more changes across the codebase, but it's more explicit. 

Still not sure which I prefer tbh.

File: components/gitpod-protocol/src/workspace-cluster.ts
Comment (andrew-farries) [reply components/gitpod-protocol/src/workspace-cluster.ts]: made manadatory in 59ad36ec515b78410261e1b9422ec09c4b7ff73e

File: components/gitpod-protocol/src/workspace-cluster.ts
Comment (andrew-farries) [reply components/gitpod-protocol/src/workspace-cluster.ts:104]: That's done in c7c3e00fc and 0abf8ea15.

File: components/ws-manager-api/typescript/src/client-provider-source.ts
Comment (andrew-farries) [reply components/ws-manager-api/typescript/src/client-provider-source.ts]: Done - assuming you meant `findByName`.

File: components/gitpod-protocol/src/workspace-cluster.ts
Comment (geropl) [inline components/gitpod-protocol/src/workspace-cluster.ts:108]: :dart:

File: components/ws-manager-api/typescript/src/client-provider-source.ts
Comment (geropl) [reply components/ws-manager-api/typescript/src/client-provider-source.ts:70]: This looks like a leftover.

**Comment by andrew-farries** [comment]
taking this out of draft as I think we are past the 'use DI/don't use DI' discussion now.

**Comment by geropl** [comment]
Code looks good, starting to test now.

**Review by geropl — APPROVED** [review]
Code LGTM, tested and works as expected. Thx @andrew-farries !

**Comment by geropl** [comment]
@andrew-farries Are we waiting for another review? Or can we start rolling it out?

**Comment by andrew-farries** [comment]
It's waiting for review form @aledbf or @sagor999

**Review by sagor999 — APPROVED** [review]

**Pull request merged**
